### ISSUE TITLE
LUA HttpClient - Add support for Keep-Alive connection

### ISSUE THREAD

**Issue by MickaelDatadome** [haproxy/haproxy#1831]
### Your Feature Request

With version 2.5 of HAProxy, a native HTTPClient has been introduced.

This client does not currently support 'Keep-Alive', and the connection is created every time.

### What are you trying to do?

Using this native HTTPClient, the goal is to propose an alternative to the SPOP.

The HTTP Client in LUA has a real performance issues when using https
- LUA with HTTPS measured 5x slower than SPOP
- LUA with HTTP measured 30% slower than SPOP

This is directly linked to the connection being closed and reopened for every single request.
A great improvement would be to handle the "Connection: Keep-Alive" header, or provide a way to keep the connection open.

### Output of `haproxy -vv`

```plain
HAProxy version 2.6.2-1ppa1~jammy 2022/07/22 - https://haproxy.org/
Status: long-term supported branch - will stop receiving fixes around Q2 2027.
Known bugs: http://www.haproxy.org/bugs/bugs-2.6.2.html
Running on: Linux 5.15.0-1017-aws #21-Ubuntu SMP Fri Aug 5 11:10:45 UTC 2022 x86_64
Build options :
  TARGET  = linux-glibc
  CPU     = generic
  CC      = cc
  CFLAGS  = -O2 -g -O2 -flto=auto -ffat-lto-objects -flto=auto -ffat-lto-objects -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2 -Wall -Wextra -Wundef -Wdeclaration-after-statement -Wfatal-errors -Wtype-limits -Wshift-negative-value -Wshift-overflow=2 -Wduplicated-cond -Wnull-dereference -fwrapv -Wno-address-of-packed-member -Wno-unused-label -Wno-sign-compare -Wno-unused-parameter -Wno-clobbered -Wno-missing-field-initializers -Wno-cast-function-type -Wno-string-plus-int -Wno-atomic-alignment
  OPTIONS = USE_PCRE2=1 USE_PCRE2_JIT=1 USE_OPENSSL=1 USE_LUA=1 USE_SLZ=1 USE_SYSTEMD=1 USE_OT=1 USE_PROMEX=1
  DEBUG   = -DDEBUG_STRICT -DDEBUG_MEMORY_POOLS

Feature list : +EPOLL -KQUEUE +NETFILTER -PCRE -PCRE_JIT +PCRE2 +PCRE2_JIT +POLL +THREAD +BACKTRACE -STATIC_PCRE -STATIC_PCRE2 +TPROXY +LINUX_TPROXY +LINUX_SPLICE +LIBCRYPT +CRYPT_H -ENGINE +GETADDRINFO +OPENSSL +LUA +ACCEPT4 -CLOSEFROM -ZLIB +SLZ +CPU_AFFINITY +TFO +NS +DL +RT -DEVICEATLAS -51DEGREES -WURFL +SYSTEMD -OBSOLETE_LINKER +PRCTL -PROCCTL +THREAD_DUMP -EVPORTS +OT -QUIC +PROMEX -MEMORY_PROFILING

Default settings :
  bufsize = 16384, maxrewrite = 1024, maxpollevents = 200

Built with multi-threading support (MAX_THREADS=64, default=1).
Built with OpenSSL version : OpenSSL 3.0.2 15 Mar 2022
Running on OpenSSL version : OpenSSL 3.0.2 15 Mar 2022
OpenSSL library supports TLS extensions : yes
OpenSSL library supports SNI : yes
OpenSSL library supports : TLSv1.0 TLSv1.1 TLSv1.2 TLSv1.3
OpenSSL providers loaded : default
Built with Lua version : Lua 5.3.6
Built with the Prometheus exporter as a service
Built with network namespace support.
Built with OpenTracing support.
Support for malloc_trim() is enabled.
Built with libslz for stateless compression.
Compression algorithms supported : identity("identity"), deflate("deflate"), raw-deflate("deflate"), gzip("gzip")
Built with transparent proxy support using: IP_TRANSPARENT IPV6_TRANSPARENT IP_FREEBIND
Built with PCRE2 version : 10.39 2021-10-29
PCRE2 library supports JIT : yes
Encrypted password support via crypt(3): yes
Built with gcc compiler version 11.2.0

Available polling systems :
      epoll : pref=300,  test result OK
       poll : pref=200,  test result OK
     select : pref=150,  test result OK
Total: 3 (3 usable), will use epoll.

Available multiplexer protocols :
(protocols marked as <default> cannot be specified using 'proto' keyword)
         h2 : mode=HTTP  side=FE|BE  mux=H2    flags=HTX|HOL_RISK|NO_UPG
       fcgi : mode=HTTP  side=BE     mux=FCGI  flags=HTX|HOL_RISK|NO_UPG
         h1 : mode=HTTP  side=FE|BE  mux=H1    flags=HTX|NO_UPG
  <default> : mode=HTTP  side=FE|BE  mux=H1    flags=HTX
       none : mode=TCP   side=FE|BE  mux=PASS  flags=NO_UPG
  <default> : mode=TCP   side=FE|BE  mux=PASS  flags=

Available services : prometheus-exporter
Available filters :
        [CACHE] cache
        [COMP] compression
        [FCGI] fcgi-app
        [  OT] opentracing
        [SPOE] spoe
        [TRACE] trace
```

**Comment by MickaelDatadome** [comment]
![image](https://user-images.githubusercontent.com/108686236/185638576-ddec90aa-ae00-4dd8-9042-2ed27613c8d0.png)
Technically possible according to sekh

**Comment by wtarreau** [comment]
I'm confused because I was going to say that there's no reason it would not work since the client uses standard muxes and outgoing connections, so I don't even see what could be the reason why connections would be closed. Once they're finished they're put back into the idle list, indexed by endpoints (src,dst,protocol). Right now the only way I'm aware of to force a connection to be closed with a server is either to *explicitely* pass "connection: close" that the mux will reinterpret as such if it wants to, or to abort in error before the end so that the connection is not reusable (only for H1, since in H2 that should only close the stream). So I'm not sure how the client manages to force a connection to be closed at the moment.

**Comment by wtarreau** [comment]
OK what I found so far is that the default server settings are not applied to the httpclient's servers, and that once the proxy is configured for "http-reuse always" it's OK:

```
diff --git a/src/http_client.c b/src/http_client.c
index cd93ade9a..165f41e0a 100644
--- a/src/http_client.c
+++ b/src/http_client.c
@@ -1130,6 +1130,7 @@ static int httpclient_precheck()
 
        proxy_preset_defaults(httpclient_proxy);
 
+       httpclient_proxy->options = (httpclient_proxy->options & ~PR_O_REUSE_MASK) | PR_O_REUSE_ALWS;
        httpclient_proxy->options |= PR_O_WREQ_BODY;
        httpclient_proxy->retry_type |= PR_RE_CONN_FAILED | PR_RE_DISCONNECTED | PR_RE_TIMEOUT;
        httpclient_proxy->options2 |= PR_O2_INDEPSTR;
@@ -1150,6 +1151,7 @@ static int httpclient_precheck()
                goto err;
        }
 
+       srv_settings_cpy(httpclient_srv_raw, &httpclient_proxy->defsrv, 0);
        httpclient_srv_raw->iweight = 0;
        httpclient_srv_raw->uweight = 0;
        httpclient_srv_raw->xprt = xprt_get(XPRT_RAW);
@@ -1169,6 +1171,7 @@ static int httpclient_precheck()
                err_code |= ERR_ALERT | ERR_FATAL;
                goto err;
        }
+       srv_settings_cpy(httpclient_srv_ssl, &httpclient_proxy->defsrv, 0);
        httpclient_srv_ssl->iweight = 0;
        httpclient_srv_ssl->uweight = 0;
        httpclient_srv_ssl->xprt = xprt_get(XPRT_SSL);
```

However this must not be merged as-is yet because I noticed that when not forcing http-reuse always, the first connection is not reused, and new ones are opened and closed, which is kind of strange. So we'd rather analyze what's happening because it could indicate that something else is going to cause problem in certain circumstances. Still checking to be certain of what we're observing, in case we spot a fix to be merged into 2.6...

**Comment by wtarreau** [comment]
Understood now. It's logical. It's just the classical chicken-and-egg problem of HTTP/1 connections in safe mode: by default we don't send new requests on existing unsafe connections (those having been used only once), and since we don't do this, we never convert these connections to safe so they are never reused.

This is not an external client, it doesn't need to be protected against reuse of a connection having seen a single request, this fixes it and in this case having http-reuse always or not doesn't change anything:
```
diff --git a/src/http_client.c b/src/http_client.c
index cd93ade9a..b500a6aad 100644
--- a/src/http_client.c
+++ b/src/http_client.c
@@ -539,6 +539,9 @@ struct appctx *httpclient_start(struct httpclient *hc)
        appctx->svcctx = hc;
        hc->flags = 0;
 
+       /* we're not an external client, no need to block reuse on first request */
+       se_fl_set(appctx->sedesc, SE_FL_NOT_FIRST);
+
        if (appctx_init(appctx) == -1) {
                ha_alert("httpclient: Failed to initialize appctx %s:%d.\n", __FUNCTION__, __LINE__);
                goto out_free_appctx;
```

We'll need to double-check all this next week, because I also managed to reproduce several times a segfault when stopping in the debugger during a close, making me think there's a race somewhere in the exit path.

**Comment by MickaelDatadome** [comment]
Waow, thanks for your feedbacks !
I'm glad my request somehow made some sense, and that you potentially found a fixable issue !

**Comment by wtarreau** [comment]
I'm just thinking about something that we could in fact improve here. If a proxy has L7 retries enabled, we could probably just ignore thet test on the first request, considering that it will be the proxy's job to retry in case of failure to send the request. I need to review all this more carefully, but that could be addressed more globally regardless of the httpclient itself.

**Comment by wtarreau** [comment]
@MickaelDatadome for your info, latest 2.7-dev now contains all the fixes for this. It would be interesting to run a test on it to measure the difference.

**Comment by wlallemand** [comment]
@MickaelDatadome were you able to test again? the fix was released with the 2.6.5 version.

**Comment by MickaelDatadome** [comment]
Thanks for the heads up, i'll try to test this next week.
Got version 2.6.5 installed ;-)
